src/template.rs

```diff
@@ -10,52 +14,81 @@ lazy_static! {
 pub struct Template {
-    src: Bytes,
-    dynamic: bool,
+    src: String,
+    src_bytes: Bytes,
+    has_ts: bool,
+    has_fields: bool,
 }
 
 impl From<&str> for Template {
     fn from(src: &str) -> Template {
         Template {
             src: src.into(),
-            dynamic: RE.is_match(src.as_bytes()),
+            src_bytes: src.into(),
+            has_ts: StrftimeItems::new(src).count() > 0,
+            has_fields: RE.is_match(src),
         }
     }
 }
 
 impl Template {
     pub fn render(&self, event: &Event) -> Result<Bytes, Vec<Atom>> {
-        if !self.dynamic {
-            return Ok(self.src.clone());
+        match (self.has_fields, self.has_ts) {
+            (false, false) => Ok(self.src_bytes.clone()),
+            (true, false) => render_fields(&self.src, event).map(Bytes::from),
+            (false, true) => Ok(render_timestamp(&self.src, event).into()),
+            (true, true) => {
+                let tmp = render_fields(&self.src, event)?;
+                Ok(render_timestamp(&tmp, event).into())
+            }
         }
+    }
 
-        let mut missing_fields = Vec::new();
-        let out = RE
-            .replace_all(self.src.as_ref(), |caps: &Captures| {
-                let key = caps
-                    .get(1)
-                    .and_then(|m| std::str::from_utf8(m.as_bytes()).ok())
-                    .map(|s| Atom::from(s))
-                    .expect("src should match regex and keys should be utf8");
-                if let Some(val) = event.as_log().get(&key) {
-                    val.to_string_lossy()
-                } else {
-                    missing_fields.push(key.clone());
-                    String::new()
-                }
-            })
-            .into_owned();
-        if missing_fields.is_empty() {
-            Ok(out.into())
-        } else {
-            Err(missing_fields)
-        }
+    pub fn render_string(&self, event: &Event) -> Result<String, Vec<Atom>> {
+        self.render(event)
+            .map(|bytes| String::from_utf8(Vec::from(bytes.as_ref())).expect("this is a bug"))
+    }
+}
+
+fn render_fields(src: &str, event: &Event) -> Result<String, Vec<Atom>> {
+    let mut missing_fields = Vec::new();
+    let out = RE
+        .replace_all(src, |caps: &Captures| {
+            let key = caps
+                .get(1)
+                .map(|s| Atom::from(s.as_str().trim()))
+                .expect("src should match regex");
+            if let Some(val) = event.as_log().get(&key) {
+                val.to_string_lossy()
+            } else {
+                missing_fields.push(key.clone());
+                String::new()
+            }
+        })
+        .into_owned();
+    if missing_fields.is_empty() {
+        Ok(out)
+    } else {
+        Err(missing_fields)
+    }
+}
+
+fn render_timestamp(src: &str, event: &Event) -> String {
+    let timestamp = match event {
+        Event::Log(log) => log.get(&event::TIMESTAMP).and_then(ValueKind::as_timestamp),
```

Comment:
minor thought: but do we want to support custom timestamp fields?

Reply:
I think we should open a separate issue for this. We don't currently give users a way to create actual custom timestamp fields/values, but when we do support that we can revisit this.